File: js/ball.js
function Ball() {
  Entity.call(this);

  this.width  = 20;
  this.height = 22;

  this.reset();
}

Ball.prototype = Object.create(Entity.prototype);
Ball.prototype.constructor = Ball;

Ball.prototype.update = function() {
  Entity.prototype.update.apply(this, arguments);

  if (this.y > game.height - this.height || this.y < 0) {
    this.yVelocity *= -1;
  }

  if (this.x > game.width) {
    game.player.score += 1;
    this.reset();
  }

  if (this.x < 0) {
    game.bot.score += 1;
    this.reset();
  }

  if (this.intersect(game.bot)) {
    var hitter = game.bot;
  } else if (this.intersect(game.player)) {
    var hitter = game.player;
  }

  if (hitter) {
    this.xVelocity *= -1.1;
    this.yVelocty *= -1.1;

    this.yVelocity += hitter.yVelocity / 2;
  }
};

Ball.prototype.reset = function() {
  this.x = game.width / 2 - this.width / 2;
  this.y = game.height / 2 - this.height / 2;

  var max = 5,
      min = -5;
  this.yVelocity = Math.floor(Math.random() * (max - min + 1) + min);
  this.xVelocity = Math.random() > 0.5 ? 5 : -5;
};

File: app/models/ball.rb
class Ball < Entity

  def initialize
    super

    @width = 20
    @height = 22

    reset
  end

  def reset
    @x = @game.width / 2 - @width / 2
    @y = @game.height / 2 - @height / 2

    max = 5
    min = -5

    @y_velocity = (rand * (max - min + 1) + min).floor
    @x_velocity = rand > 0.5 ? 5 : -5
  end

  def update
    super

    if (@y > @game.height - @height || y < 0)
      @y_velocity *= -1
    end

    if @x > @game.width
      @game.player.score += 1
      reset
    end

    if @x < 0
      @game.bot.score += 1
      reset
    end

    if intersect(@game.bot)
      hitter = @game.bot
    elsif intersect(@game.player)
      hitter = @game.player
    end

    if hitter
      @x_velocity *= -1.1
      @y_velocty *= -1.1

      @y_velocity += hitter.y_velocity / 2
    end
  end

end
